394 enhance engines to provide unified workflow

Nic-Ma opened the pull request:
Fixes #394 .

### Description
This PR enhanced our `engines` module with `trainers` and `evaluators`.

They can provide **unified** and **easier** training and evaluation workflows, it will be useful to set advanced strategies, like: JSON config, AutoML and Federated Learning, etc.

For next steps, I will also do these items in other PRs:
1. Update and enhance our handlers, add more useful handlers.
2. Develop workflow integration test.

### Status
**Ready**

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or new feature that would cause existing functionality to change)
- [ ] New tests added to cover the changes
- [x] Docstrings/Documentation updated

Nic-Ma commented:
/black

Nic-Ma commented:
/black

ericspod commented on monai/engines/workflow.py:
```diff
@@ -0,0 +1,121 @@
+from .utils import default_prepare_batch
+
+
+class Workflow(ABC):
```
I'd suggest this class should inherit `ignite.engine.Engine` instead of having an internally created instance of it. This would simplify code so that `self.engine` wouldn't be needed everywhere, and any existing code expecting `Engine` instances will function just by passing in a `Workflow` instance without having to type check.

Nic-Ma replied:
updated to inherit Engine in the latest patch.
Thanks.

ericspod commented on monai/engines/workflow.py:
```diff
@@ -0,0 +1,121 @@
+
+    """
+
+    def __init__(
```
It would be nice to have an optional callable parameter which is used as the step function for the `Engine`, if not given `self._iteration` is used as the default.

Nic-Ma replied:
added step_function in the latest patch.
Thanks.

ericspod commented on monai/engines/workflow.py:
```diff
@@ -0,0 +1,121 @@
+    ):
+        # FIXME:
+        if amp:
+            print("Will add AMP support when PyTorch v1.6 released.")
```
This should be a log message.

Nic-Ma replied:
updated in the latest patch.
Thanks.

ericspod commented on monai/engines/workflow.py:
```diff
@@ -0,0 +1,121 @@
+        # FIXME:
+        if amp:
+            print("Will add AMP support when PyTorch v1.6 released.")
+        assert isinstance(device, torch.device), "must provide PyTorch device information."
```
Instead of assert, a conditional should be used to validate inputs which raise `ValueError` if the condition isn't met.

Nic-Ma replied:
updated in the latest patch.
Thanks.

ericspod commented on monai/engines/workflow.py:
```diff
@@ -0,0 +1,121 @@
+                        engine.state.best_metric_epoch = engine.state.epoch
+
+        if handlers is not None and len(handlers) > 0:
+            assert isinstance(handlers, (list, tuple)), "handlers must be a chain."
```
If `handlers` is not iterable the for loop below will raise an exception so this check isn't strictly necessary and doesn't permit other types besides list or tuple.

Nic-Ma replied:
removed the assert in the latest patch.
Thanks.

ericspod commented:
There is some discussion with Ignite on high level API types that has overlap with this PR: https://github.com/pytorch/ignite/issues/912

Nic-Ma commented:
> There is some discussion with Ignite on high level API types that has overlap with this PR: [pytorch/ignite#912](https://github.com/pytorch/ignite/issues/912)

Hi @ericspod ,

Thanks very much for your review and sharing!
I will totally adjust the workflows tomorrow based on your previous investigation and comments.
Maybe we can use this PR as the baseline and then enhance & add more workflows later.
I think the unified training workflow will be very useful, especially for JSON/YAML config, AutoML and Federated Learning, etc.
Thanks.

vfdev-5 commented:
Thanks for mentioning @wyli ! Yes, definitely, classes like `SupervisedTrainer` make a lot of sense and it is something we aim to provide in Ignite too while keeping the flexibility of Engine.

PS: I also cc @sdesrozis as @justusschock is not working anymore on Ignite.

Nic-Ma commented on monai/engines/utils.py:
```diff
@@ -0,0 +1,53 @@
+
+class CommonKeys:
+    IMAGE = "image"
+    LABEL = "label"
```
updated in the latest patch.
Thanks.

Nic-Ma commented:
/black

Nic-Ma commented:
Hi @ericspod and @wyli ,

Thanks very much for your review and feedback!
I updated the PR according to your comments.
Could you please help review it again?
Thanks.

sdesrozis commented on monai/handlers/validation_handler.py:
```diff
@@ -0,0 +1,41 @@
+from ignite.engine import Events
+
+
+class ValidationHander:
```
Typo error ? `ValidationHandler`

Nic-Ma replied:
updated in the latest patch.
Thanks.

sdesrozis commented on tests/test_handler_validation.py:
```diff
@@ -0,0 +1,47 @@
+        # set up testing handler
+        val_data_loader = torch.utils.data.DataLoader(Dataset(data))
+        evaluator = TestEvaluator(torch.device("cpu:0"), val_data_loader)
+        saver = ValidationHander(evaluator, interval=2)
```
Typo again about ˋValidationHandler` ?

Nic-Ma replied:
updated in the latest patch.
Thanks.

sdesrozis commented:
Hi ! I reviewed (quite late) this PR because closely related to ignite API. I really appreciate the smart usage of ignite done here. it makes me think that ignite is exactly in the right position for projects like MONAI : providing building blocks to compose complex applications.

My feedback about the PR is maybe we need in ignite a way to define smart functions in `Engine`, a bit more than a `Callable` to help about `amp` for instance. Moreover, I think that the work we are currently doing about agnostic distributed computing (for tpu, cuda, etc.) should be helpful for MONAI. We will support on that way if you need 👍🏻

Nic-Ma commented:
Hi @sdesrozis ,

Thanks very much for your review and feedback from ignite side, glad to co-work with you guys.
I updated the PR according to your comments.
And we will add AMP support when PyTorch v1.6 released(it's just reserved for API now).
We may leverage the AMP feature in ignite directly.

Thanks.

Nic-Ma commented:
Hi @ericspod and @wyli ,

I updated this PR with all the latest comments, could you please help review it again?
Thanks.

Nic-Ma commented:
/black

Nic-Ma commented:
Thanks for @wyli 's review, I updated according to his offline comments:
1. change `_run()` to `run()`.
2. delete the `_iteration()` abstract function in Trainer and Evaluator.
3. add doc-string for the `global_epoch` parameter of Evaluator.
4. change `RegularInferer` to `SimpleInferer`.
5. fix several small typos in doc-strings and documentation.
6. refine all the doc-strings in this PR.

I also changed all the `assert` to `raise Exception` according to @ericspod 's comments.

@ericspod , @wyli , Could you please help review it again?
Thanks in advance.

wyli commented on monai/engines/evaluator.py:
```diff
@@ -0,0 +1,141 @@
+            device,
+            1,
+            False,
+            val_data_loader,
```
could use `device=device, max_epochs=1, amp=False, ... ` for code readability (same for the trainers)

Nic-Ma commented:
/black

Nic-Ma commented:
Thanks @wyli for your review!

Hi @ericspod ,

Do you have any other comments or suggestions?
Thanks.

ericspod commented:
It looks good in general. The approach I had taken with a higher level `Engine` type was to have a base type inheriting `Engine` which served the role of an inferer as well as being subtyped as a `Trainer` and `Evaluator` classes to suite those roles. The idea was that this provided a simple inference capability regardless of which subtype was used, so one could have inference routines that could work with either. 

Another idea was to have a method implementing the network forward pass (pass arguments to the network, packaging results into a tuple) which both subtypes would use, this would reduce code and allow adapting the network to data parallel operations without creating a wrapper type. Similarly there was a method for wrapping the forward pass for the loss function. The idea with this was to allow a subtype to modify how the forward passes worked without having to change any of the other code. 

I'm fine with merging this now though, these are ideas for evolution into the future perhaps.

Nic-Ma commented:
Hi @ericspod ,

Thanks for your review and for sharing cool ideas!
I think your 2 proposals are very interesting, they are mainly focusing on the optimization for the operations of `inferer` and `network`or `loss`, which are not enhanced in this PR. This PR is mainly to enhance the operations for `iteration`, `handlers`, and `metrics` in the ignite engines. 
So maybe we can work together later to **combine** your proposals with current Trainer and Evaluator. I will merge this PR first and develop other higher level modules in later PRs.

Thanks.
